subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java

```diff
@@ -282,23 +280,62 @@ private Provider<JavaLauncher> toolchainLauncher(Project project) {
         return service.launcherFor(extension.getToolchain());
     }
 
+    /**
+     * A state shared by the validation process.
+     * The state is first collected by the {@link Jar#filesMatching(String, Action)} during the bundling.
+     * Then {@link PluginValidationAction} (added to the Jar as doLast) processes the collected information to emit warnings if necessary.
+     * <p>
+     * List and Set cannot be shared without this wrapper class as their identity isn't preserved by the configuration cache.
```

Comment:
This makes sense as a description of your design for this change, but not as much as explaining the resulting code.

Instead, I would say use more general terms like wrapping the state into this class is required so it remains a shared object when storing it into/loading it from the configuration cache (collections in are duplicated when serialized).